lectures/zksnark.tex

```diff
@@ -0,0 +1,408 @@
+
+How can we translate an if statement?
+
+\begin{lstlisting}[language=C,numbers=none]
+    if (a) {
+        return b * c
+    } else {
+        return b + c
+    }
+\end{lstlisting}
+
+We can express this logic in mathematical terms as follows: "If $a$ is true, compute 
+$b \times c$; otherwise, compute $b + c$." Only numerical expressions are allowed. Knowing that
+$\text{true} := 1$ and $\text{false} := 0$, we can transform it as follows:
+\[ r = a \times (b \times c) + (1 - a) \times (b + c) \]
+Where the witness vector is: $w = (r, a , b, c)$: $(6, 1, 2, 3)$, $(5, 0, 2, 3)$.\\
+
+But, we need to verify all the intermediate steps. This can be achieved by transforming the above
+equation using the simplest terms (the gates), ensuring the correctness of each step in the program: 
+
+\begin{equation*}
+    \begin{aligned}
+        r_1 &= b \times c \\
+        r_2 &= b + c \\
+        r_3 &= 1 - a \\
+        r_4 &= a \times r_1 \\
+        r_5 &= r_3 \times r_2 \\
+        r &= r_4 + r_5 \\
+    \end{aligned}
+\end{equation*}
+
+With witness vector: $w = (r, r_1, r_2, r_3, r_4, r_5, a , b, c)$: $(6, 6, 5, 0, 6, 0, 1, 2, 3)$.
+
+\subsection{Rank-1 Constraint System}
+
+Almost any program written in high-level programming language can be translated (compiled) into
+arithmetic circuits, that are really powerfull tool. But for the ZK proof we need slightly different
+format of it - \textbf{Rank-1 Constrain System}, where the simpliest term is \textbf{constraint}. 
+This offers a more flexible and general way to describe these parts.
+    
+\begin{definition}
+    Let $\mathbb{V}$ be a vector space over the field $\mathbb{F}$. An \textbf{inner product} 
+    on $\mathbb{V}$ is a function:
+    \[ \<\cdot, \cdot \>: \mathbb{V} \times \mathbb{V} \rightarrow \mathbb{F} \]
+    defined for $u, v \in \mathbb{V}$ as a linear combination:
+    \[ \< u, v \> = \sum_{i=1}^{n} u_i v_i \]
+    With next properties:
+    \begin{itemize}
+        \item Symmetry: $\forall u, v \in \mathbb{V}: \< u, v \> = \overline{\< v, u \>}$
+        \item Linearity in the first argument: $\forall u, v, w \in \mathbb{V}, \forall a \in 
+        \mathbb{F}: \< au + v, w \> = a \< u, w \> + \< v, w \>$
+        \item Positivity: $\forall u \in \mathbb{V}: \< u, u \> \geq 0$ and 
+        $\< u, u \> = 0 \Leftrightarrow u = 0$
+    \end{itemize}
+\end{definition}
+
+
+\begin{example}
+    Let $u, v$ are vectors over the real number $\mathbb{R}$, where
+    \[u = (1, 2, 3), v = (2, 4, 3)\], then:
+    $\<u, v\> = \sum_{i=1}^{3}u_iv_i = 2*1 + 2*4 + 3*3 = 2 + 8 + 9 = 19$
+\end{example}
+
+With knowledge of the inner product of two vectors, we can now formulate a definition of the 
+constraint in the context of an R1CS.
+
+\begin{definition}
+    Each \textbf{constraint} in the Rank-1 Constraint System must be in the form:
+    \[ \<a, w\> \times \<b, w\> = \<c, w\> \]
+    Where $\mathbf{w}$ is a vector containing all the \textit{input}, \textit{output}, and 
```

Comment:
Make notation consistent. If you do not make any special notation for a vector (like, $w$), do not use `\mathbf{...}` here. Otherwise, use this `\mathbf{...}` ($\mathbf{w}$) everywhere (which I do recommend). Alternatively, quite frequently, one denotes vectors using `\vec{...}` to put an arrow above ($\vec{w}$).